### DISCUSSION TITLE
Plan for July/Aug - following the release of wlroots 0.18

### DISCUSSION THREAD

**Discussion by johanmalm** [labwc/labwc#1979] [General]
Following a `labwc` release we normally go into a merge window. For the next couple of weeks I particularly had in mind PRs
- #1956
- #1965
- #1941
- #1962
- #1963
- \+ a bunch of stuff from the issues/discussions.

However... `wlroots 0.18` was also released yesterday and as a result I think we should prioritise porting to 0.18 (#1641) and putting other PRs on hold.

There is stuff left in:
- https://github.com/labwc/labwc/pull/1641#issue-2193900273
- https://github.com/labwc/labwc/pull/1641#issuecomment-2106333963

...but following some IRC discussion it feels like the only thing we ought to fix before merge is:

- "decorations" to avoid the "xdg configure before surface initialize" warning
- sort commits + get the proper wlroots commits against each one

I think the advantages of this approach are that it gets 0.18 out for wider testing early in our release cycle; and it's likely to be easier to do the final chasing against the master branch - particularly if we get feedback on applications misbehaving as we'll then have 'test cases' and can track breaking changes more easily.

Appreciate thoughts and feedback on this proposal.

Cc: @ahesford @Consolatis @jlindgren90 @tokyo4j @jp7677

**Comment by tokyo4j** [comment]
I agree with your proposal, though I'm not sure how much I can help.
IMO not introducing regressions is more important than fixing existing bugs.

**Comment by ahesford** [comment]
Pushing the wlroots update to the front of the queue is OK with me.

**Comment by Consolatis** [comment]
Just a general remark: now that the wlroots headers are versioned, distros can easily install multiple version of wlroots simultaneously. So we are not in a rush and can basically sail the wlroots 0.17.x sailboat for as long as we want.

Having the chase branch either be part of the labwc repo itself (currently its still a branch in my GH fork) or having it merged in the master branch makes follow up PRs easier though. For example [tearing without the NO_ATOMIC](https://aur.archlinux.org/cgit/aur.git/tree/labwc-atomic-modesetting-tearing.patch?h=labwc-wlroots-018-git) requirement.

**Reply by kode54** [reply]
I'll definitely file that PR when things are sorted.

**Comment by kode54** [comment]
Here's another tentative patch / MR candidate, I'm seeking testing for this as well.

https://aur.archlinux.org/cgit/aur.git/tree/0001-wip-handle-renderer-loss.patch?h=labwc-wlroots-018-git

It does manage to recover the display, but doesn't seem to be perfect at restoring every output or every output's windows/resources without a full off/on cycle for each output, which it doesn't do yet. Wondering if I'm doing something wrong here.

**Reply by johanmalm** [reply]
Thanks

**Reply by johanmalm** [reply]
Consider adding `output_config_apply()` here too.

https://github.com/labwc/labwc/blob/3be20dc6c71f3de07fe7e1a9e3b729a9710e69fd/src/output.c#L471

**Reply by kode54** [reply]
And where do I get the config from?

**Reply by johanmalm** [reply]
Ah, sorry, my advice was ill thought through. We don't store a copy of internal state for outputs.
It would be a pain to have to do that, but if it's the only way, then let's do it.
What is it that's actually lost on GPU reset? Do we lose everything (like width, height, refresh rate, scale, transform) or just some specifics?

**Reply by kode54** [reply]
The problem I had was with the DRM subsystem, not with the recovery of the compositor. It failed to reset whatever firmware was needed for adaptive sync, so my GPU now blanks the screen when the refresh rate is being reduced. I filed a bug report with drm/amd on freedesktop.org.

Edit: And I worked around it configuration-wise by disabling adaptiveSync in config and disabling it on that output. I'll see if the feature comes back after a reboot, whenever I happen to reboot again.

**Comment by johanmalm** [comment]
Thanks for feedback. The more I look at this, the more I'm convinced that early merge of 0.18 is the better option :smile:

I've had another look at https://github.com/labwc/labwc/pull/1641#issuecomment-2106333963 and we may wish to copy items 4,5,6 (relating to sway implementation of layer-shell popups) at some point, but quite frankly I'd rather do that in slow time after the merge as they're pretty small and I'd rather change as little as possible at each step. I've tested with gtk3-layer-demo, xfce4-panel, sfwbar, lxqt-panel (Application Menu and Fancy Application) menu and they all work as expected.

Similarly with GPU resets (both the `reconfigure` for server buffers and the @kode54 patch above) I'd be comfortable with picking up after the merge on the basis that they are improvements and I think it'll then be easier to bisect + fault-find.

All in all I think it becomes false economy to hang out for perfection on a branch that gets little exposure / test-coverage.

**Reply by kode54** [reply]
I added a reconfigure to my recovery patch, it's needed to recover all of the server side decorations, at least.

**Comment by droc12345** [comment]
I have no preference for merging order, but I do believe that wlroots 0.18 is going to be needed, as any improvements in
wlroots will have 0.18 as a base, with new work happening on 0.19-dev. Some things will get backported to 0.17 but not everything.
